pkg/controllers/management/cis/scheduled_scan.go

```diff
@@ -0,0 +1,281 @@
+				clusterInfo.cronSchedule = cluster.Spec.ScheduledClusterScan.ScheduleConfig.CronSchedule
+				logrus.Infof("scheduledScanHandler: clusterSync: cron schedule modified for cluster %v: %v",
+					cluster.Name, clusterInfo.cronSchedule)
+			}
+		}
+	} else {
+		if ok {
+			logrus.Infof("scheduledScanHandler: clusterSync: Scheduled Scan disabled for cluster: %v",
+				cluster.Name)
+			delete(ssh.clustersMap, cluster.Name)
+		}
+	}
+	return nil, nil
+}
+
+func (ssh *scheduledScanHandler) sync(key string, _ *v3.ClusterScan) (runtime.Object, error) {
+	if !strings.HasPrefix(key, ScheduledScanKeyPrefix) {
+		return nil, nil
+	}
+	splits := strings.Split(strings.TrimPrefix(key, ScheduledScanKeyPrefix), separator)
+	clusterID := splits[0]
+	pastCronSchedule := splits[1]
+	ssh.Lock()
+	clusterInfo, ok := ssh.clustersMap[clusterID]
+	ssh.Unlock()
+	if !ok {
+		logrus.Debugf("scheduledScanHandler: sync: scheduled scan has been disabled, ignoring")
+		return nil, nil
+	}
+	cluster, err := ssh.clusterLister.Get("", clusterID)
+	if err != nil {
+		return nil, fmt.Errorf("scheduledScanHandler: sync: error fetching cluster %v: %v", clusterID, err)
+	}
+
+	if !cluster.Spec.ScheduledClusterScan.Enabled {
+		logrus.Debugf("scheduledScanHandler: sync: scheduled scan is disabled, ignoring the current one")
+		return nil, nil
+	}
+
+	if err := cis.ValidateClusterBeforeLaunchingScan(cluster); err != nil {
+		return nil, err
+	}
+
+	currentCronSchedule := getCronSchedule(cluster)
+	if currentCronSchedule != pastCronSchedule {
+		logrus.Infof("scheduledScanHandler: sync: cron schedule has changed for cluster: %v, hence skipping",
+			clusterID)
+		return nil, nil
+	}
+
+	if _, err = checkAndLaunchScan(cluster, ssh.clusterClient, ssh.clusterScanClient, clusterInfo); err != nil {
+		return nil, fmt.Errorf("scheduledScanHandler: clusterSync: cluster %v: error launching scan: %v",
+			cluster.Name, err)
+	}
+
+	retention := getRetention(cluster)
+	clusterScanClient := ssh.managementClient.ClusterScans(clusterID)
+	if err := cleanOldScans(clusterID, retention, clusterScanClient, ssh.clusterScanLister); err != nil {
+		return nil, fmt.Errorf("scheduledScanHandler: clusterSync: error cleaning old scans: %v", err)
+	}
+
+	return nil, nil
+}
+
+func checkAndLaunchScan(cluster *v3.Cluster,
+	clusterClient v3.ClusterInterface,
+	clusterScanClient v3.ClusterScanInterface,
+	clusterInfo *scheduleInfo,
+) (*v3.ClusterScan, error) {
+	var cisScanConfig *v3.CisScanConfig
+	if cluster.Spec.ScheduledClusterScan.ScanConfig == nil ||
+		cluster.Spec.ScheduledClusterScan.ScanConfig.CisScanConfig == nil {
+		cisScanConfig = &v3.CisScanConfig{
+			Profile: v3.CisScanProfileTypePermissive,
+		}
+	}
+
+	nextAt := clusterInfo.nextScanAt
+	now := time.Now().Round(time.Second)
+	if !nextAt.Equal(now) && !now.After(nextAt) {
+		logrus.Debugf("scheduledScanHandler: checkAndLaunchScan: cluster %v next: %v, now: %v",
+			cluster.Name, nextAt.String(), now.String())
+		return nil, nil
```

Comment:
when you return from here, the next scan is not scheduled, should't we be scheduling the next one to maintain the schedule?

Reply:
Oh I see that you have added !nextAt.Equal(now) && !now.After(nextAt) - With now >= nextAt, this looks good.